### DISCUSSION TITLE
Mobile Observability

### DISCUSSION THREAD

**Discussion by jingwei99** [MobileNativeFoundation/discussions#39] [General]
At Lyft, currently we use analytics events for mobile observability. It’s a pretty standard setup for mobile apps.

The analytics events get fired when something happens, such as user interactions, app life cycle events and etc. Different metadata can be attached to each analytics event.

Analytics events are later consumed in two ways: 1) SQL queries for complex analysis; 2) near real-time dashboards for monitoring and automated alarms.

Meanwhile, recently we started integrating a new reporting mechanism in Lyft apps, it utilizes [Envoy Mobile](https://github.com/envoyproxy/envoy-mobile)’s stats client [Pulse](https://eng.lyft.com/pulse-stats-apis-from-envoy-mobile-6db71c6a2f22) to send lightweight stats aiming to be more real-time.

Would love to see how other apps approach the topic of mobile observability. 

Just throwing out a few questions: what’s the setup for the app you are working/worked on? What’s the learning? What are the things you found interesting/useful? What are your thoughts on real-time observability for mobile, and how would you utilize it for your app.

**Comment by sgleadow** [comment]
Mobile observability is definitely an area where many companies seem to be immature, compared with the set up I see for backend systems. We had a discussion early last year, in the beforetimes, with local mobile companies near our office in Melbourne, Australia and testing analytics and just general observability were common painpoints, so I'm glad you've brought this up and can't want to hear what other people are doing!

At realestate.com.au, we have fairly mature log aggregation and New Relic integration on our server systems, and use this to drive most of our operational support. Actual usage of products, both web and native apps, rely more on analytics for this as you suggest (both dashboard and via SQL).

We're starting to use New Relic Mobile and Browser more heavily in our front ends as well - not just for crashes, which feels like bare minimum app observability, but also custom events and using this to drive more advanced altering and operational support. We could (and do) use analytics to fill some of this gap, but in my mind, they're trying to fulfil different purposes: the needs of a product team understanding user behaviour are very different from a platform team wanting to understand the technical behaviour.

I wouldn't say we're mature in our use of New Relic Mobile in our apps yet, but it's something we want to invest in. It's not the cheapest tool around, but it feels great if you're in between a small company (maybe too expensive?) or really large companies (end up hand rolling custom a custom solution like @jingwei99 references with Envoy proxy above).

**Comment by andre-alves** [comment]
Besides monitoring crashes with Bugsnag, here at ifood.com.br we also use two other systems for observability.

We have analytics events, more focused on the user and product interactions. These events are sent to a data lake for further SQL queries and complex analysis. At the moment, this system is only D-1 and we are unable to create real-time dashboards because of that, but it's a future wish to solve that.

Another complementary and simpler system that we are exploring is to send logs and metrics in a standardized format to an ELK stack (we currently use Logz.io). The simplicity of this data makes aggregation fast and easy to create some dashboards and real-time alerts monitoring the health of the app, all at a low cost.

**Reply by sgleadow** [reply]
Thanks @andre-alves - the remote logging is definitely something on my mind. We can do a bit of that with custom New Relic events, but it's not really the same as what we get on the server with ELK stacks (we use Splunk for aggregation and search across all our backend systems).

On our web systems, we use a combination of New Relic and Splunk, since the server-rendered components allow for easier log aggregation.

Neither web or our mobile apps do any real remote logging for us to observe though. We do (or did, I'd need to check) keep local logs in a file on the phone that users could attach to support requests so we could help debug what was going on. This didn't really provide generic observability but was still useful without the ongoing network cost of remote logging.

**Reply by tarunanand-dev** [reply]
Andre - thanks. Did you roll a customer ELK client for android/iOS? How does it work? or did Logz.io provide those clients. I cant seem to find any good open source clients.

**Reply by andre-alves** [reply]
@tarunanand-dev Logz.io doesn't provide mobile clients AFAIK. We just use their HTTPS API with JSON and gzip:
https://docs.logz.io/shipping/log-sources/json-uploads.html

**Comment by lucas1295santos** [comment]
One issue that I've faced before, is how to create **useful** logging data. In a way that it is easy to analyze it later and actually understand what is going on. Sometimes we find that a problem is happening, but we can't figure out what exactly by analyzing the logs, so we are in the dark when trying to fix the issue, as we're not sure of the root of the problem.

I'd love to hear how this problem is tackled in other apps. Do you have a standard structure? Do you analyze ahead on what kind of information you will need to understand a possible issue later?

**Comment by contrecc** [comment]
Getting meaningful logging data from a mobile app. At Embrace, we work with many great mobile companies and have seen several successful ways they go about adding logs that are helpful and not just noise. 

Mobile operating systems are complicated, and they are only getting more complex.  Still, the ways in which they can crash are somewhat bounded by the operating system and hardware.

With the apps we've worked with, the most useful logging is when it either helps categorize a problem or learn how to reproduce it. This leads to some natural categories to log:

1) Any entry and exit points, log how the user got to where they are. Is this the process launch?  A notification?  A location-based wakeup?  This becomes the bookends of the troubleshooting process.  Remember that things like Permission dialogues are often missed exit points from the app, they can open settings.

2) All major navigation events should be logged, view changes, button presses, keyboard presentation/dismissal are all great candidates.  The goal is to log as sparse a map as possible of the user journey, while still retaining enough to reproduce what they were doing.

3) From here the developers can start to zero in on the major crash categories:
        3a) Memory, log the basic wired page values to get an idea of memory at the time of the crash/log
        3b) CPU State, is it at 100% cpu, or 0% -- these yield different troubleshooting steps, deadlock vs hanging
        3c) Thread ID and priority, are we on the main thread or not -- should this function run off the main thread?

These basic items provide a good foundation for useful logging in an app.  From here you would build into domain-specific things that your particular app is concerned with.

**Comment by pyricau** [comment]
I can describe Square, on the point of sale side (a set of apps sharing a lot of core code). I have more experience with the Android side but a lot here applies to iOS as well.

We've found that there are different categories of events and needs. In the past we've all mixed them together but we're starting to rethink our approach a bit. We've built many different solutions over time, not always very consistently

# Product analytics

- Goal: Understand customer behavior and how a feature is used, and also triggers.
- Target: PMs
- Upload: batches
- Destination: all sorts of places. Data lake, triggers for sending emails, etc.

We use something internal that is somewhat similar to Segment.io . In theory that makes it easy to enrich the data, consume it by different parties in clear ways, etc. This is a fairly new pipeline for us (before we were sending all those logs to the same place as the technical analytics)

# Technical analytics

- Goal: monitor the health of the app, and identify improvement targets
- Target: Engs + EMs (and PMs when the data is high level enough)
- Upload: batches
- Destination: Data lake (sql queryable) usually with 24h delay due to volume, and a long retention (years)

Some examples / details

- All network requests send by the client are logged as events, with client side durations, trace ids, status.. or logged as a network error. The trace id is super useful to then find the request in our server side trace UI. That data is incredibly useful to figure out the real performance of endpoints, whether we need to bring services closer to mobile devices based on region, etc.
- Every log includes a UUID that's unique per process. We also include that id in our bugsnag crashes, which enables us to go from a crash on bugsnag to all technical events for that app process lifetime and vice versa.
- The data generally turns into a dashboard, and maybe alerts and monitoring. There's also quite a lot of ad hoc querying as we're investigating issues.
- Our perf data also goes there, as individual events (e.g. startup, navigation with frame rendering duration, etc). I think we'll evolve our story around perf data over time.

# Crashes

- Goal: Keep crash rate low by controlled app rollout and provide rich data for investigation
- Target: Engs + Release Ops
- Upload: immediately on each event / stored on disk & uploaded async if no network
- Destination: bugsnag then imported into a data lake to build custom crash metrics and join with other technical data.

Our release ops team manages branch cuts, coordinates with QA, gates release fixes from engs, and decides when to start & halt roll outs, primarily based on crash rate. Once a top crash surfaces, they typically file a JIRA and ping engs. We enrich our crash reports with a ton of data client side, which helps reproducing and fixing.

# Diagnostic zip upload

This is sort of a one off we built a while ago for our custom hardware. When troubleshooting issues with Customer Success reps, they might ask a customer to go into a special section of the app to click a button that will trigger a zip upload with a ton of data that helps understand what's been happening on device. In a way this is somewhat similar to the OS bug reports on Android, though it's not the same data at all. This doesn't scale at all, and there's a ton of logs to sift through, but it's helpful when investigating an oncall where we're completely in the dark.

# Debug logs

Debug logs are specifically to investigate issues happening to a specific customer or set of customers, and we want to understand what's going on with their device. Usually you want to look at one device, sift through a ton of logs and start spotting patterns, unexpected logs, etc. 

We used to send semi structured debug logs to the same place as the technical analytics and product analytics. It would end up in a data lake we could query with SQL, and that was really nice because you could pick a device and see everything that had happening to it (no PII of course) to help debug an issue. However, that creates data volume issues. We want long retention for analytics, but not for debug logs.

Currently we're designing the next system for this, which might be based on Elastic Search (with a good frontend). Retention will probably be ~30 days.

# Heap analysis

We're experimenting with heap analysis (Android) in released apps. It's still very early and we don't have a good backend for receiving the data. More here: https://square.github.io/leakcanary/changelog/#leakcanary-for-releases
